cohortextractor/validate_dummy_data.py

```diff
@@ -0,0 +1,131 @@
+from datetime import datetime
+
+import pandas as pd
+
+
+SUPPORTED_FILE_FORMATS = ["csv", "csv.gz"]
+
+
+class DummyDataValidationError(Exception):
+    pass
+
+
+def validate_dummy_data(column_definitions, dummy_data_file):
+    """Validate that dummy data provided by user matches expected structure and format.
+
+    Raises DummyDataValidationError if dummy data is not valid.
+    """
+
+    df = read_into_dataframe(dummy_data_file)
+    validate_expected_columns(df, column_definitions)
+
+    for col_name, query_node in column_definitions.items():
+        if col_name == "population":
+            # Ignore the population definition, since it is not used as a column in the
+            # output
+            continue
+
+        validator = get_csv_validator(col_name, query_node)
+
+        for ix, value in enumerate(df[col_name]):
+            if pd.isna(value) or not value:
+                # Ignore null or missing values
+                continue
+
+            try:
+                validator(value)
+            except ValueError:
+                raise DummyDataValidationError(
+                    f"Invalid value `{value!r}` for {col_name} in row {ix + 2}"
+                )
+
+    for ix, value in enumerate(df["patient_id"]):
+        if not isinstance(value, int):
+            raise DummyDataValidationError(
+                f"Invalid value `{value!r}` for patient_id in row {ix + 2}"
+            )
+
+
+def read_into_dataframe(path):
+    """Read data from path into a Pandas DataFrame."""
+
+    try:
+        if path.suffixes in [[".csv"], [".csv", ".gz"]]:
```

Comment:
Could we remove the duplication between this and `SUPPORTED_FILE_FORMATS`?

Reply:
Eventually there will be more supported file formats, and it won't look so duplicated.  It's like this in the old CE:
```
        if path.suffixes in [[".csv"], [".csv", ".gz"]]:
            return pd.read_csv(path)
        elif path.suffixes in [[".dta"], [".dta", ".gz"]]:
            return pd.read_stata(path)
        elif path.suffix == ".feather":
            return pd.read_feather(path)
        else:
            msg = f"Dummy data must be in one of the following formats: {', '.join(SUPPORTED_FILE_FORMATS)}"
            raise DummyDataValidationError(msg)
```

I can change it to 
```
suffixes = ".".join([suffix.strip(".") for suffix in path.suffixes])
if suffixes in SUPPORTED_FILE_FORMATS:
    return pd.read_csv(path)
```

But presumably later we'll have different file extensions to deal with and we'll need something more like this again

Reply:
Even then, I think we'd want to somehow pull the list of extensions together with the handling code to help us check they're not inconsistent. Just putting the definitions next to each other would help.